### PR TITLE
feat: adding custom field of type PHONE

### LINKED ISSUE #5556: [CAL-407] Better phone number field
I would be good to have a more intuitive field for phone number.

Where you can choose the country, the number length based on country. Also the possiblity to pre-populate this field on URL

> From [SyncLinear.com](https://synclinear.com)

### PR DESCRIPTION
**Pull request by Udit-takkar**
Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Fixes: https://github.com/calcom/cal.com/issues/5556

https://user-images.githubusercontent.com/53316345/203270839-3b5ff72a-0c2a-40ee-a4fc-6119fc2fbee3.mov

### CODE REVIEW

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 1]
```diff
@@ -299,6 +299,46 @@ const BookingPage = ({
...
+    const bookingCustomInputs = Object.keys(booking.customInputs || {}).map((inputId) => ({
+      label: eventType.customInputs.find((input) => input.id === parseInt(inputId))?.label || "",
+      value:
+        booking.customInputs && inputId in booking.customInputs && booking.customInputs[inputId]
+          ? booking.customInputs[inputId]
+          : "",
+    }));
+
+    // Check if custom input of type Phone number is valid
+    const customInputsWithPhoneNumberType = eventType.customInputs.filter(
```
Comment (Udit-takkar) [inline apps/web/components/booking/pages/BookingPage.tsx:311]: validating phone number in custom input only. rest are already done by zod.

**Comment by PeerRich** [comment]
remember to translate error messages too

**Comment by Udit-takkar** [comment]
the visual error message is translated https://github.com/calcom/cal.com/blob/99179404f26a9f80a1b1403ece8bd1cc926e6c89/apps/web/components/booking/pages/BookingPage.tsx#L796

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ import {
   WebhookTriggerEvents,
 } from "@prisma/client";
 import async from "async";
+import { isValidPhoneNumber } from "libphonenumber-js";
```
Comment (zomars) [inline packages/features/bookings/lib/handleNewBooking.ts:10]: Let's add `"libphonenumber-js": "^1.10.12",` to `packages/features/package.json`

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 3]
```diff
@@ -294,6 +294,46 @@ const BookingPage = ({
   }
 
   const bookEvent = (booking: BookingFormValues) => {
+    const bookingCustomInputs = Object.keys(booking.customInputs || {}).map((inputId) => ({
```
Comment (zomars) [inline apps/web/components/booking/pages/BookingPage.tsx:297]: We're traversing `booking.customInputs` once

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 4]
```diff
@@ -294,6 +294,46 @@ const BookingPage = ({
...
+      label: eventType.customInputs.find((input) => input.id === parseInt(inputId))?.label || "",
```
Comment (zomars) [inline apps/web/components/booking/pages/BookingPage.tsx:298]: Here we're traversing  `booking.customInputs` again to find an input.

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 5]
```diff
@@ -294,6 +294,46 @@ const BookingPage = ({
...
+      value:
+        booking.customInputs && inputId in booking.customInputs && booking.customInputs[inputId]
+          ? booking.customInputs[inputId]
+          : "",
+    }));
+
+    // Check if custom input of type Phone number is valid
+    const customInputsWithPhoneNumberType = eventType.customInputs.filter(
```
Comment (zomars) [inline apps/web/components/booking/pages/BookingPage.tsx:306]: Here we're traversing `booking.customInputs` again to filter phone inputs.

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 6]
```diff
@@ -294,6 +294,46 @@ const BookingPage = ({
...
+      (customInput) => customInput.type === EventTypeCustomInputType.PHONE
+    );
+
+    if (customInputsWithPhoneNumberType.length) {
+      let isErrorFound = false;
+      customInputsWithPhoneNumberType.forEach((customInput) => {
```
Comment (zomars) [inline apps/web/components/booking/pages/BookingPage.tsx:312]: Here we're traversing filtered phone `booking.customInputs` to validate each one.

**Review by zomars — COMMENTED** [review]
Code seems to work but I think we can achieve the same result with less traversal (maybe we can use `reduce` to process `booking.customInputs` in one go.

File: apps/web/components/booking/pages/BookingPage.tsx (hunk 4 above)
Comment (sean-brydon) [reply apps/web/components/booking/pages/BookingPage.tsx:301]: Hopefully soon we can iradicate this behaviour when we move the BookingCustomInputs away from a json field and into an actual table.

This was discussed in https://github.com/calcom/cal.com/discussions/5840

**Comment by Udit-takkar** [comment]
> Code seems to work but I think we can achieve the same result with less traversal (maybe we can use reduce to process booking.customInputs in one go.

@zomars i fixed the rest of the things you mentioned and for less traversal thing i discussed this with sean. he said  this approach is fine until we move to an extra table.

**Comment by sean-brydon** [comment]
> > Code seems to work but I think we can achieve the same result with less traversal (maybe we can use reduce to process booking.customInputs in one go.
> 
> @zomars i fixed the rest of the things you mentioned and for less traversal thing i discussed this with sean. he said this approach is fine until we move to an extra table.

On the back of this @zomars this traversal is how we have always done it - it has just been moved closer to scope (from what i can see anyway)

**Review by zomars — APPROVED** [review]
Thanks you @Udit-takkar Looks mergeable now

Ship it

**Pull request merged**